src/Compilers/Core/Portable/Symbols/CompilerFeatureRequiredHelpers.cs

```diff
@@ -0,0 +1,81 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Reflection.Metadata;
+
+namespace Microsoft.CodeAnalysis;
+
+[Flags]
+internal enum CompilerFeatureRequiredFeatures
+{
+    None = 0,
+    RefStructs = 1 << 0,
+    RequiredMembers = 1 << 1,
+
+    Unknown = 1 << 31
+}
+
+internal static class CompilerFeatureRequiredHelpers
+{
+    internal static string? GetUnsupportedCompilerFeature(EntityHandle token, PEModule containingModule, CompilerFeatureRequiredFeatures allowedFeatures)
+    {
+        if (token.IsNil)
+        {
+            return null;
+        }
+
+        Debug.Assert((allowedFeatures & CompilerFeatureRequiredFeatures.Unknown) == 0);
+
+        var compilerRequiredFeatures = containingModule.GetCompilerFeatureRequiredFeatures(token);
+
+        if (tryGetFirstUnsupportedFeature(compilerRequiredFeatures, out var unsupportedFeature))
+        {
+            return unsupportedFeature;
+        }
+
+        // Check the containing module and assembly as well, if the symbol itself was fine
```

Comment:
>// Check the containing module and assembly as well, if the symbol itself was fine [](http://example.com/codeflow?start=8&length=82)

Are we skipping declaring method and type?

Reply:
> Are we skipping declaring method and type?

Yes. They will get checked on their own, it's only assembly and module symbols that don't have their own use-site diagnostics.

Reply:
> They will get checked on their own, 

Are you certain? If we are consuming a static method, are we going to check type?

Reply:
> If we are consuming a static method, are we going to check the declaring type? 

Yes. I added a variant for that case to be sure.

Reply:
Note: this didn't prove true for VB, as it removes static receivers, while C# keeps them in the tree. I added an extra part to the VB side to check the containing type.